File: backend/app/api/incidents.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from backend.app.core.database import get_db

from backend.app.schemas.incident import (
    IncidentCreate,
    IncidentResponse,
)

from backend.app.services.audit_service import AuditService
from backend.app.services.approval_service import ApprovalService
from backend.app.services.context_service import ContextService
from backend.app.services.detection_service import DetectionService
from backend.app.services.incident_service import IncidentService
from backend.app.services.rca_service import RCAService
from backend.app.services.remediation_service import RemediationService


router = APIRouter(
    prefix="/api/v1/incidents",
    tags=["Incidents"],
)


@router.post(
    "",
    response_model=IncidentResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_incident(
    data: IncidentCreate,
    db: Session = Depends(get_db),
):
    return IncidentService.create_incident(db, data)


@router.get(
    "",
    response_model=list[IncidentResponse],
)
def get_incidents(
    db: Session = Depends(get_db),
):
    return IncidentService.get_incidents(db)


@router.post(
    "/detect",
    response_model=list[IncidentResponse],
)
def detect_incidents(
    db: Session = Depends(get_db),
):
    return DetectionService.detect(db)


@router.get(
    "/{incident_id}/analysis",
)
def analyze_incident(
    incident_id: str,
    db: Session = Depends(get_db),
):
    analysis = RCAService.analyze(
        db,
        incident_id,
    )

    if analysis is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Incident not found",
        )

    return analysis


@router.get(
    "/{incident_id}/context",
)
def get_incident_context(
    incident_id: str,
    db: Session = Depends(get_db),
):
    context = ContextService.collect(
        db,
        incident_id,
    )

    if context is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Incident not found",
        )

    return context


@router.get(
    "/{incident_id}/remediation",
)
def get_remediation_plan(
    incident_id: str,
    db: Session = Depends(get_db),
):
    plan = RemediationService.plan(
        db,
        incident_id,
    )

    if plan is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Incident not found",
        )

    return plan


@router.post(
    "/{incident_id}/approve",
)
def approve_remediation(
    incident_id: str,
    db: Session = Depends(get_db),
):
    result = ApprovalService.approve(
        db,
        incident_id,
    )

    if result is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Incident not found",
        )

    return result


@router.post(
    "/{incident_id}/execute",
)
def execute_remediation(
    incident_id: str,
    db: Session = Depends(get_db),
):
    result = ApprovalService.execute(
        db,
        incident_id,
    )

    if result is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Incident not found",
        )

    return result


@router.post(
    "/{incident_id}/rollback",
)
def rollback_remediation(
    incident_id: str,
    db: Session = Depends(get_db),
):
    result = ApprovalService.rollback(
        db,
        incident_id,
    )

    if result is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Incident not found",
        )

    return result


@router.get(
    "/{incident_id}/audit",
)
def get_incident_audit(
    incident_id: str,
    db: Session = Depends(get_db),
):
    logs = AuditService.get_incident_logs(
        db,
        incident_id,
    )

    return [
        {
            "id": log.id,
            "incident_id": log.incident_id,
            "action": log.action,
            "status": log.status,
            "details": log.details,
            "created_at": log.created_at.isoformat(),
        }
        for log in logs
    ]


@router.get(
    "/{incident_id}",
    response_model=IncidentResponse,
)
def get_incident(
    incident_id: str,
    db: Session = Depends(get_db),
):
    incident = IncidentService.get_incident(
        db,
        incident_id,
    )

    if incident is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Incident not found",
        )

    return incident

File: frontend/src/api/incidents.js
import axios from "axios";


// ============================================================
// API CONFIGURATION
// ============================================================
const API_BASE_URL = "https://ai-sre.onrender.com";

const api = axios.create({
  baseURL: API_BASE_URL,
  timeout: 30000,

  headers: {
    "Content-Type": "application/json",
  },
});


// ============================================================
// REQUEST ERROR HANDLER
// ============================================================

function handleApiError(error) {
  console.error("AI-SRE API Error:", error);

  if (error.response) {
    console.error(
      "Status:",
      error.response.status
    );

    console.error(
      "Response:",
      error.response.data
    );
  } else if (error.request) {
    console.error(
      "No response received from backend."
    );
  } else {
    console.error(
      "Request error:",
      error.message
    );
  }

  throw error;
}


// ============================================================
// INCIDENTS
// ============================================================

export async function getIncidents() {
  try {
    const response = await api.get(
      "/api/v1/incidents"
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


export async function getIncident(
  incidentId
) {
  try {
    const response = await api.get(
      `/api/v1/incidents/${incidentId}`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// AUTOMATIC DETECTION
// ============================================================

export async function detectIncidents() {
  try {
    const response = await api.post(
      "/api/v1/incidents/detect"
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// AI ANALYSIS
// ============================================================

export async function analyzeIncident(
  incidentId
) {
  try {
    const response = await api.get(
      `/api/v1/incidents/${incidentId}/analysis`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// INCIDENT CONTEXT
// ============================================================

export async function getIncidentContext(
  incidentId
) {
  try {
    const response = await api.get(
      `/api/v1/incidents/${incidentId}/context`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// REMEDIATION
// ============================================================

export async function getRemediation(
  incidentId
) {
  try {
    const response = await api.get(
      `/api/v1/incidents/${incidentId}/remediation`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// APPROVE REMEDIATION
// ============================================================

export async function approveRemediation(
  incidentId
) {
  try {
    const response = await api.post(
      `/api/v1/incidents/${incidentId}/approve`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// EXECUTE REMEDIATION
// ============================================================

export async function executeRemediation(
  incidentId
) {
  try {
    const response = await api.post(
      `/api/v1/incidents/${incidentId}/execute`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// ROLLBACK REMEDIATION
// ============================================================

export async function rollbackRemediation(
  incidentId
) {
  try {
    const response = await api.post(
      `/api/v1/incidents/${incidentId}/rollback`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// AUDIT LOGS
// ============================================================

export async function getAuditLogs(
  incidentId
) {
  try {
    const response = await api.get(
      `/api/v1/incidents/${incidentId}/audit`
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// TELEMETRY - METRICS
// ============================================================

export async function createMetric(
  metric
) {
  try {
    const response = await api.post(
      "/api/v1/telemetry/metrics",
      metric
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


export async function getMetrics(
  serviceId = null
) {
  try {
    const url = serviceId
      ? `/api/v1/telemetry/metrics?service_id=${encodeURIComponent(
          serviceId
        )}`
      : "/api/v1/telemetry/metrics";

    const response = await api.get(url);

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// TELEMETRY - LOGS
// ============================================================

export async function createLog(
  log
) {
  try {
    const response = await api.post(
      "/api/v1/telemetry/logs",
      log
    );

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


export async function getLogs(
  serviceId = null
) {
  try {
    const url = serviceId
      ? `/api/v1/telemetry/logs?service_id=${encodeURIComponent(
          serviceId
        )}`
      : "/api/v1/telemetry/logs";

    const response = await api.get(url);

    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}


// ============================================================
// EXPORT API INSTANCE
// ============================================================

export default api;